tools/fast/bin/localnet/main.go

```diff
@@ -0,0 +1,350 @@
+			return
+		}
+	}
+
+	if ok, err := isEmpty(workdir); !ok {
+		if err == nil {
+			err = fmt.Errorf("workdir is not empty")
+		}
+
+		exitcode = handleError(err, "fail when checking workdir;")
+		return
+	}
+
+	env, err := fast.NewEnvironmentMemoryGenesis(&balance, workdir)
+	if err != nil {
+		exitcode = handleError(err)
+		return
+	}
+
+	// Defer the teardown, this will shuteverything down for us
+	defer env.Teardown(ctx) // nolint: errcheck
+
+	binpath, err := testhelpers.GetFilecoinBinary()
+	if err != nil {
+		exitcode = handleError(err, "no binary was found, please build go-filecoin;")
+		return
+	}
+
+	// Setup localfilecoin plugin options
+	options := make(map[string]string)
+	options[lpfc.AttrLogJSON] = "0"                                     // Disable JSON logs
+	options[lpfc.AttrLogLevel] = "4"                                    // Set log level to Info
+	options[lpfc.AttrUseSmallSectors] = fmt.Sprintf("%t", smallSectors) // Enable small sectors
+	options[lpfc.AttrFilecoinBinary] = binpath                          // Use the repo binary
+
+	genesisURI := env.GenesisCar()
+	genesisMiner, err := env.GenesisMiner()
+	if err != nil {
+		exitcode = handleError(err, "failed to retrieve miner information from genesis;")
+		return
+	}
+
+	fastenvOpts := fast.EnvironmentOpts{
+		InitOpts:   []fast.ProcessInitOption{fast.POGenesisFile(genesisURI)},
+		DaemonOpts: []fast.ProcessDaemonOption{fast.POBlockTime(series.GlobalSleepDelay)},
+	}
+
+	// The genesis process is the filecoin node that loads the miner that is
+	// define with power in the genesis block, and the prefunnded wallet
+	genesis, err := env.NewProcess(ctx, lpfc.PluginName, options, fastenvOpts)
+	if err != nil {
+		exitcode = handleError(err, "failed to create genesis process;")
+		return
+	}
+
+	err = series.SetupGenesisNode(ctx, genesis, genesisMiner.Address, files.NewReaderFile(genesisMiner.Owner))
+	if err != nil {
+		exitcode = handleError(err, "failed series.SetupGenesisNode;")
+		return
+	}
+
+	// Create the processes that we will use to become miners
+	var miners []*fast.Filecoin
+	for i := 0; i < count; i++ {
+		miner, err := env.NewProcess(ctx, lpfc.PluginName, options, fastenvOpts)
+		if err != nil {
+			exitcode = handleError(err, "failed to create miner process;")
+			return
+		}
+
+		miners = append(miners, miner)
+	}
+
+	// We will now go through the process of creating miners
+	// InitAndStart
+	// 1. Initialize node
+	// 2. Start daemon
+	//
+	// Connect
+	// 3. Connect to genesis
+	//
+	// SendFilecoinDefaults
+	// 4. Issue FIL to node
+	//
+	// CreateMinerWithAsk
+	// 5. Create a new miner
+	// 6. Set the miner price, and get ask
+	//
+	// ImportAndStore
+	// 7. Generated some random data and import it to genesis
+	// 8. Genesis propposes a storage deal with miner
+	//
+	// WaitForDealState
+	// 9. Query deal till posted
+
+	var deals []*storagedeal.Response
+
+	for _, miner := range miners {
+		err = series.InitAndStart(ctx, miner)
```

Comment:
Unsure if this is the correct place for this comment, but we will need to shut these daemons down at some point. As things currently stand the daemons will continue to run in the background until the user manually kills them.

Reply:
Only if the user kills the process before the setup completes. This seems to be due to the fact that defer functions are not run after Ctrl-C.